sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx

```diff
@@ -0,0 +1,140 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=True
+
+""" For internal use only. No backwards compatibility guarantees."""
+
+cimport cython
+from libc.stdint cimport int64_t, INT64_MAX
+from libc.stdlib cimport malloc, free
+
+cdef extern from "stdlib.h" nogil:
+  void *memset(void *str, int c, size_t n)
+
+cdef int64_t* MAX_LONG_10_FOR_LEADING_ZEROS = [19, 18, 18, 18, 18, 17, 17, 17,
+                                               16, 16, 16, 15,15, 15, 15, 14,
+                                               14, 14, 13, 13, 13, 12, 12, 12,
+                                               12, 11, 11, 11, 10, 10, 10, 9, 9,
+                                               9, 9, 8, 8,8, 7, 7, 7, 6, 6, 6,
+                                               6, 5, 5, 5, 4, 4, 4, 3, 3, 3, 3,
+                                               2, 2, 2, 1, 1, 1, 0, 0, 0]
+
+cdef unsigned long long* POWER_TEN = [10e-1, 10e0, 10e1, 10e2, 10e3, 10e4, 10e5,
+                                      10e6, 10e7, 10e8, 10e9, 10e10, 10e11,
+                                      10e12, 10e13, 10e14, 10e15, 10e16, 10e17,
+                                      10e18]
+
+
+cdef inline bint compare_to(int64_t x, int64_t y):
+  """return the sign bit of x-y"""
+  if x < y:
+    return 1
+  return 0
+
+
+cdef int64_t bit_length(int64_t element):
+  """Same function as python built_in int.bit_length"""
+  cdef int64_t bit_count = 0
+  while element > 0:
+    bit_count += 1
+    element >>= 1
+  return bit_count
+
+
+cdef int64_t get_log10_round_to_floor(int64_t element):
```

Comment:
This code is a bit hard to follow. Given you have a list of powers of ten, it might make sense to just do

```
int power = 1
while element < POWER_TEN[power]:
    power += 1
return power - 1
```